### PR TITLE
Fmegen/supporting from endpoint with token

### PR DESCRIPTION
**Pull request by fmegen**

### CODE REVIEW

File: src/sdk/SpeechConfig.ts [hunk 1]
```diff
@@ -39,10 +39,15 @@ export abstract class SpeechConfig {
     }
 
     /**
-     * Creates an instance of the speech factory with specified endpoint and subscription key.
-     * This method is intended only for users who use a non-standard service endpoint or paramters.
-     * the language setting in uri takes precedence, and the effective language is "de-DE".
+     * Creates an instance of the speech config with specified endpoint and subscription key.
+     * This method is intended only for users who use a non-standard service endpoint or parameters.
      * Note: Please use your LanguageUnderstanding subscription key in case you want to use the Intent recognizer.
+     * Note: The query parameters specified in the endpoint URL are not changed, even if they are set by any other APIs.
```
Comment (zhouwangzw) [inline src/sdk/SpeechConfig.ts]: Is this tested?

Comment (fmegen) [reply src/sdk/SpeechConfig.ts]: nope, updated the docs for now.

File: src/sdk/SpeechTranslationConfig.ts [hunk 2]
```diff
@@ -43,6 +43,9 @@ export abstract class SpeechTranslationConfig extends SpeechConfig {
      *       expires, the caller needs to refresh it by setting the property authorizationToken with a new
      *       valid token. Otherwise, all the recognizers created by this SpeechTranslationConfig instance
      *      will encounter errors during recognition.
+     * As configuration values are copied when creating a new recognizer, the new token value will not apply to recognizers that have already been created.
```
Comment (zhouwangzw) [inline src/sdk/SpeechTranslationConfig.ts]: Indentation?

File: src/sdk/SpeechConfig.ts [hunk 3]
```diff
@@ -39,10 +39,15 @@ export abstract class SpeechConfig {
...
+     * For example, if language is defined in the uri as query parameter "language=de-DE", and also set by CreateSpeechRecognizer("en-US"),
```
Comment (mahilleb-msft) [inline src/sdk/SpeechConfig.ts]: >CreateSpeechRecognizer("en-US") [](start = 107, length = 31)

bad ref?

File: src/sdk/SpeechConfig.ts [hunk 4]
```diff
@@ -39,10 +39,15 @@ export abstract class SpeechConfig {
...
+     * the language setting in uri takes precedence, and the effective language is "de-DE".
+     * Only the parameters that are not specified in the endpoint URL can be set by other APIs.
+     * Note: To use authorization token with fromEndpoint, pass an empty string to the subscriptionKey in the fromEndpoint method,
+     * and then call setAuthorizationToken() on the created SpeechConfig instance to use the authorization token.
```
Comment (mahilleb-msft) [inline src/sdk/SpeechConfig.ts]: >and then call setAuthorizationToken( [](start = 7, length = 36)

bad ref? is there a property?

File: src/sdk/Recognizer.ts [hunk 5]
```diff
@@ -186,7 +186,7 @@ export abstract class Recognizer {
                 new Context(new OS(osPlatform, osName, osVersion))));
 
         const subscriptionKey = this.privProperties.getProperty(PropertyId.SpeechServiceConnection_Key, undefined);
-        const authentication = subscriptionKey ?
+        const authentication = (subscriptionKey && subscriptionKey !== "") ?
             new CognitiveSubscriptionKeyAuthentication(subscriptionKey) :
             new CognitiveTokenAuthentication(
                 (authFetchEventId: string): Promise<string> => {
```
Comment (zhouwangzw) [inline src/sdk/Recognizer.ts]: just curios: why do we need to do this twice?

File: src/sdk/SpeechTranslationConfig.ts [hunk 6]
```diff
@@ -87,8 +92,8 @@ export abstract class SpeechTranslationConfig extends SpeechConfig {
 
     /**
      * Gets/Sets the authorization token.
-     * If this is set, subscription key is ignored.
-     * User needs to make sure the provided authorization token is valid and not expired.
+     * Note: The caller needs to ensure that the authorization token is valid. Before the authorization token
+     * expires, the caller needs to refresh it by calling this setter with a new valid token.
```
Comment (mahilleb-msft) [inline src/sdk/SpeechTranslationConfig.ts]: >calling this setter [](start = 50, length = 19)

setting this property? there's no setter

File: src/sdk/SpeechTranslationConfig.ts [hunk 7]
```diff
@@ -68,6 +71,8 @@ export abstract class SpeechTranslationConfig extends SpeechConfig {
      *       "language=de-DE", and also set by the speechRecognitionLanguage property, the language
      *       setting in uri takes precedence, and the effective language is "de-DE".
      * Only the properties that are not specified in the endpoint URL can be set by other APIs.
+     * Note: To use authorization token with fromEndpoint, pass an empty string to the subscriptionKey in the fromEndpoint method,
+     * and then call setAuthorizationToken() on the created SpeechConfig instance to use the authorization token.
```
Comment (mahilleb-msft) [inline src/sdk/SpeechTranslationConfig.ts]: >setAuthorizationToken() [](start = 21, length = 23)

cf. elsewhere. if there's a real property and no setter, should be changed

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts [hunk 8]
```diff
@@ -208,3 +208,179 @@ test("AuthToken refresh works correctly", (done: jest.DoneCallback) => {
...
+        objsToClose.push(s);
+
+        // now set the authentication token
+        s.authorizationToken = authToken;
+
+        let pumpSilence: boolean = false;
+        let bytesSent: number = 0;
+        let streamStopped: boolean = false;
+
+        // Pump the audio from the wave file specified with 1 second silence between iterations indefinetly.
```
Comment (mahilleb-msft) [inline tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: >indefinetly [](start = 96, length = 11)

typo

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts [hunk 9]
```diff
@@ -208,3 +208,179 @@ test("AuthToken refresh works correctly", (done: jest.DoneCallback) => {
...
+test("AuthToken refresh works even with fromEndpoint configurations", (done: jest.DoneCallback) => {
+    // tslint:disable-next-line:no-console
+    console.info("Name: AuthToken refresh works correctly");
+
+    if (!Settings.ExecuteLongRunningTestsBool) {
+        // tslint:disable-next-line:no-console
+        console.info("Skipping test.");
+        done();
+        return;
+    }
```
Comment (mahilleb-msft) [inline tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: this added test is not running in Azure DevOps?

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts [hunk 10]
```diff
@@ -208,3 +208,179 @@ test("AuthToken refresh works correctly", (done: jest.DoneCallback) => {
...
+    WaitForCondition(() => {
+        return !!authToken;
+    }, () => {
+        // Pump valid speech and then silence until at least one speech end cycle hits.
+        const fileBuffer: ArrayBuffer = WaveFileAudioInput.LoadArrayFromFile(Settings.WaveFile);
+
+        // make a connection to the standard speech endpoint.
+        const endpointString: string = "wss://" + Settings.SpeechRegion + ".stt.speech.microsoft.com";
+
+        const s: sdk.SpeechConfig = sdk.SpeechConfig.fromEndpoint(new URL(endpointString), Settings.SpeechRegion);
```
Comment (wolfma61) [inline tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: gets the endpoint and the region?

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts [hunk 11]
```diff
@@ -208,3 +208,179 @@ test("AuthToken refresh works correctly", (done: jest.DoneCallback) => {
             });
     });
 }, 1000 * 60 * 25); // 25 minutes.
+
+test("AuthToken refresh works even with fromEndpoint configurations", (done: jest.DoneCallback) => {
```
Comment (zhouwangzw) [inline tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: Does it test emty subscription key with fromEndpoint?

File: src/sdk/SpeechConfig.ts (hunk 1 above)
Comment (fmegen) [reply src/sdk/SpeechConfig.ts]: checked.

File: src/sdk/SpeechTranslationConfig.ts (hunk 2 above)
Comment (fmegen) [reply src/sdk/SpeechTranslationConfig.ts]: this was on purpose since this is a separate topic

Comment (fmegen) [reply src/sdk/SpeechTranslationConfig.ts]: fixed

File: src/sdk/SpeechConfig.ts (hunk 3 above)
Comment (fmegen) [reply src/sdk/SpeechConfig.ts]: fixed

File: src/sdk/Recognizer.ts (hunk 5 above)
Comment (fmegen) [reply src/sdk/Recognizer.ts]: yes, the wonder of js comparism rules

File: src/sdk/SpeechConfig.ts (hunk 4 above)
Comment (fmegen) [reply src/sdk/SpeechConfig.ts]: fixed

File: src/sdk/SpeechTranslationConfig.ts (hunk 6 above)
Comment (fmegen) [reply src/sdk/SpeechTranslationConfig.ts]: SpeechTranslationConfig extends SpeechConfig and speechconfig has a getter.

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts (hunk 8 above)
Comment (fmegen) [reply tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: fixed

File: tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts (hunk 9 above)
Comment (fmegen) [reply tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: i hope it is part of the end-2-end testing since it takes >10minutes

Comment (fmegen) [reply tests/LongRunning/SpeechRecoAuthTokenRefreshTests.ts]: @mahilleb-msft  is there a way to specify this in the pullrequest?

File: src/sdk/SpeechTranslationConfig.ts (hunk 7 above)
Comment (fmegen) [reply src/sdk/SpeechTranslationConfig.ts]: fixed

File: src/sdk/Recognizer.ts (hunk 5 above)
Comment (fmegen) [reply src/sdk/Recognizer.ts]: that is from the previous lib implementation where there was a separation of first and subsequent retrievals

---
In reply to: [258072439](https://github.com/Microsoft/cognitive-services-speech-sdk-js/pull/36#discussion_r258072439) [](ancestors = 258072439)

**Pull request merged**